{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n",
      "<utilities.VirtualModel object at 0x7fb2543ea470> 137.73148148148147\n",
      "<utilities.VirtualModel object at 0x7fb253896b00> 136.57407407407408\n",
      "<utilities.VirtualModel object at 0x7fb253896b38> 154.96399176954733\n",
      "<utilities.VirtualModel object at 0x7fb253896eb8> 145.83333333333331\n",
      "<utilities.VirtualModel object at 0x7fb253896ef0> 251.54320987654322\n",
      "<utilities.VirtualModel object at 0x7fb2538482b0> 69.83024691358025\n",
      "<utilities.VirtualModel object at 0x7fb2538482e8> 134.64506172839506\n",
      "<utilities.VirtualModel object at 0x7fb253848668> 124.4212962962963\n",
      "<utilities.VirtualModel object at 0x7fb2538486a0> 137.92438271604937\n",
      "<utilities.VirtualModel object at 0x7fb253848a20> 131.55864197530866\n",
      "<utilities.VirtualModel object at 0x7fb253848da0> 152.77777777777777\n",
      "<utilities.VirtualModel object at 0x7fb253854160> 112.01131687242798\n",
      "attempting to recover from pickled file\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/LEMS_2007One.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "{'injected_square_current': {'amplitude': array(346.64351851851853) * pA, 'delay': array(100.0) * ms, 'duration': array(1000.0) * ms}}\n",
      "[ 0.005] s\n",
      "[ 723.4025] s\n",
      "200 200\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/LEMS_2007One.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "{'injected_square_current': {'amplitude': array(8.294753086419753) * pA, 'delay': array(100.0) * ms, 'duration': array(1000.0) * ms}}\n",
      "[ 0.005] s\n",
      "[ 770.1925] s\n",
      "200 200\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/LEMS_2007One.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "{'injected_square_current': {'amplitude': array(56.90586419753086) * pA, 'delay': array(100.0) * ms, 'duration': array(1000.0) * ms}}\n",
      "[ 0.005] s\n",
      "[ 824.63] s\n",
      "200 200\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/LEMS_2007One.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "{'injected_square_current': {'amplitude': array(152.2633744855967) * pA, 'delay': array(100.0) * ms, 'duration': array(1000.0) * ms}}\n",
      "[ 0.005] s\n",
      "[ 648.5275] s\n",
      "200 200\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/LEMS_2007One.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "{'injected_square_current': {'amplitude': array(280.60699588477365) * pA, 'delay': array(100.0) * ms, 'duration': array(1000.0) * ms}}\n",
      "[ 0.005] s\n",
      "[ 842.0225] s\n",
      "200 200\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/LEMS_2007One.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "{'injected_square_current': {'amplitude': array(3.8580246913580245) * pA, 'delay': array(100.0) * ms, 'duration': array(1000.0) * ms}}\n",
      "[ 0.005] s\n",
      "[ 647.28] s\n",
      "200 200\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/LEMS_2007One.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "/home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2/x86_64 gets here?\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/scratch/other/neuronunit/neuronunit/optimization/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "{'injected_square_current': {'amplitude': array(36.07253086419753) * pA, 'delay': array(100.0) * ms, 'duration': array(1000.0) * ms}}\n",
      "[ 0.005] s\n",
      "[ 685.01] s\n",
      "200 200\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'plt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-9ab038bf6ccc>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    353\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    354\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 355\u001b[1;33m \u001b[0mplt\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'plt' is not defined"
     ]
    }
   ],
   "source": [
    "from IPython.lib.deepreload import reload\n",
    "import ipyparallel as ipp\n",
    "rc = ipp.Client(profile='default')\n",
    "rc[:].use_cloudpickle()\n",
    "dview = rc[:]\n",
    "\n",
    "import pickle\n",
    "import deap\n",
    "from deap import base\n",
    "from deap import creator\n",
    "%load_ext autoreload\n",
    "%autoreload\n",
    "import net_graph\n",
    "\n",
    "toolbox = base.Toolbox()\n",
    "import utilities\n",
    "\n",
    "\n",
    "def update_vm_pop(pop, trans_dict):\n",
    "    '''\n",
    "    inputs a population of genes/alleles, the population size MU, and an optional argument of a rheobase value guess\n",
    "    outputs a population of genes/alleles, a population of individual object shells, ie a pickleable container for gene attributes.\n",
    "    Rationale, not every gene value will result in a model for which rheobase is found, in which case that gene is discarded, however to\n",
    "    compensate for losses in gene population size, more gene samples must be tested for a successful return from a rheobase search.\n",
    "    If the tests return are successful these new sampled individuals are appended to the population, and then their attributes are mapped onto\n",
    "    corresponding virtual model objects.\n",
    "    '''\n",
    "    from itertools import repeat\n",
    "    import numpy as np\n",
    "    import copy\n",
    "    pop = [toolbox.clone(i) for i in pop ]\n",
    "    #import utilities\n",
    "    def transform(ind):\n",
    "        '''\n",
    "        Re instanting Virtual Model at every update vmpop\n",
    "        is Noneifying its score attribute, and possibly causing a\n",
    "        performance bottle neck.\n",
    "        '''\n",
    "        vm = utilities.VirtualModel()\n",
    "\n",
    "        param_dict = {}\n",
    "        for i,j in enumerate(ind):\n",
    "            param_dict[trans_dict[i]] = str(j)\n",
    "        vm.attrs = param_dict\n",
    "        vm.name = vm.attrs\n",
    "        vm.evaluated = False\n",
    "        return vm\n",
    "\n",
    "\n",
    "    if len(pop) > 0:\n",
    "        vmpop = dview.map_sync(transform, pop)\n",
    "        vmpop = list(copy.copy(vmpop))\n",
    "    else:\n",
    "        # In this case pop is not really a population but an individual\n",
    "        # but parsimony of naming variables\n",
    "        # suggests not to change the variable name to reflect this.\n",
    "        vmpop = transform(pop)\n",
    "    return vmpop\n",
    "\n",
    "\n",
    "\n",
    "def check_rheobase(vmpop,pop=None):\n",
    "    '''\n",
    "    inputs a population of genes/alleles, the population size MU, and an optional argument of a rheobase value guess\n",
    "    outputs a population of genes/alleles, a population of individual object shells, ie a pickleable container for gene attributes.\n",
    "    Rationale, not every gene value will result in a model for which rheobase is found, in which case that gene is discarded, however to\n",
    "    compensate for losses in gene population size, more gene samples must be tested for a successful return from a rheobase search.\n",
    "    If the tests return are successful these new sampled individuals are appended to the population, and then their attributes are mapped onto\n",
    "    corresponding virtual model objects.\n",
    "    '''\n",
    "    def check_fix_range(vms):\n",
    "        '''\n",
    "        Inputs: lookup, A dictionary of previous current injection values\n",
    "        used to search rheobase\n",
    "        Outputs: A boolean to indicate if the correct rheobase current was found\n",
    "        and a dictionary containing the range of values used.\n",
    "        If rheobase was actually found then rather returning a boolean and a dictionary,\n",
    "        instead logical True, and the rheobase current is returned.\n",
    "        given a dictionary of rheobase search values, use that\n",
    "        dictionary as input for a subsequent search.\n",
    "        '''\n",
    "        import pdb\n",
    "        import copy\n",
    "        import numpy as np\n",
    "        import quantities as pq\n",
    "        sub=[]\n",
    "        supra=[]\n",
    "        steps=[]\n",
    "        vms.rheobase=0.0\n",
    "        for k,v in vms.lookup.items():\n",
    "            if v==1:\n",
    "                #A logical flag is returned to indicate that rheobase was found.\n",
    "                vms.rheobase=float(k)\n",
    "                vms.steps = 0.0\n",
    "                vms.boolean = True\n",
    "                return vms\n",
    "            elif v==0:\n",
    "                sub.append(k)\n",
    "            elif v>0:\n",
    "                supra.append(k)\n",
    "\n",
    "        sub=np.array(sub)\n",
    "        supra=np.array(supra)\n",
    "\n",
    "        if len(sub)!=0 and len(supra)!=0:\n",
    "            #this assertion would only be wrong if there was a bug\n",
    "            print(str(bool(sub.max()>supra.min())))\n",
    "            assert not sub.max()>supra.min()\n",
    "        if len(sub) and len(supra):\n",
    "            everything = np.concatenate((sub,supra))\n",
    "\n",
    "            center = np.linspace(sub.max(),supra.min(),7.0)\n",
    "            centerl = list(center)\n",
    "            # The following code block probably looks counter intuitive.\n",
    "            # Its job is to delete duplicated search values.\n",
    "            # Ie everything is a list of everything already explored.\n",
    "            # It then makes a corrected center position.\n",
    "            for i,j in enumerate(centerl):\n",
    "                if i in list(everything):\n",
    "\n",
    "                    np.delete(center,i)\n",
    "                    del centerl[i]\n",
    "                    # delete the duplicated elements element, and replace it with a corrected\n",
    "                    # center below.\n",
    "            #delete the index\n",
    "            #np.delete(center,np.where(everything is in center))\n",
    "            #make sure that element 4 in a seven element vector\n",
    "            #is exactly half way between sub.max() and supra.min()\n",
    "            center[int(len(center)/2)+1]=(sub.max()+supra.min())/2.0\n",
    "            steps = [ i*pq.pA for i in center ]\n",
    "\n",
    "        elif len(sub):\n",
    "            steps = np.linspace(sub.max(),2*sub.max(),7.0)\n",
    "            np.delete(steps,np.array(sub))\n",
    "            steps = [ i*pq.pA for i in steps ]\n",
    "\n",
    "        elif len(supra):\n",
    "            steps = np.linspace(-2*(supra.min()),supra.min(),7.0)\n",
    "            np.delete(steps,np.array(supra))\n",
    "            steps = [ i*pq.pA for i in steps ]\n",
    "\n",
    "        vms.steps = steps\n",
    "        vms.rheobase = None\n",
    "        return copy.copy(vms)\n",
    "\n",
    "\n",
    "    def check_current(ampl,vm):\n",
    "        '''\n",
    "        Inputs are an amplitude to test and a virtual model\n",
    "        output is an virtual model with an updated dictionary.\n",
    "        '''\n",
    "\n",
    "        global model\n",
    "        import quantities as pq\n",
    "        import get_neab\n",
    "        from neuronunit.models import backends\n",
    "        from neuronunit.models.reduced import ReducedModel\n",
    "        from neuronunit.models import backends\n",
    "        from neuronunit.models.reduced import ReducedModel\n",
    "        import get_neab\n",
    "        import os\n",
    "        print(get_neab.LEMS_MODEL_PATH)\n",
    "        new_file_path = '{0}{1}'.format(str(get_neab.LEMS_MODEL_PATH),int(os.getpid()))\n",
    "        print(new_file_path)\n",
    "\n",
    "        os.system('cp ' + str(get_neab.LEMS_MODEL_PATH)+str(' ') + new_file_path)\n",
    "        model = ReducedModel(new_file_path,name='vanilla',backend='NEURON')\n",
    "        model.load_model()\n",
    "        #new_file_path = str(get_neab.LEMS_MODEL_PATH)+str(int(os.getpid()))\n",
    "        #model = ReducedModel(new_file_path,name=str('vanilla'),backend='NEURON')\n",
    "        #model.load_model()\n",
    "        model.update_run_params(vm.attrs)\n",
    "\n",
    "        DELAY = 100.0*pq.ms\n",
    "        DURATION = 1000.0*pq.ms\n",
    "        params = {'injected_square_current':\n",
    "                  {'amplitude':100.0*pq.pA, 'delay':DELAY, 'duration':DURATION}}\n",
    "\n",
    "\n",
    "        if float(ampl) not in vm.lookup or len(vm.lookup)==0:\n",
    "\n",
    "            current = params.copy()['injected_square_current']\n",
    "\n",
    "            uc = {'amplitude':ampl}\n",
    "            current.update(uc)\n",
    "            current = {'injected_square_current':current}\n",
    "            vm.run_number += 1\n",
    "            model.update_run_params(vm.attrs)\n",
    "            model.inject_square_current(current)\n",
    "            vm.previous = ampl\n",
    "            n_spikes = model.get_spike_count()\n",
    "            vm.lookup[float(ampl)] = n_spikes\n",
    "            if n_spikes == 1:\n",
    "                vm.rheobase = float(ampl)\n",
    "\n",
    "                vm.name = str('rheobase {0} parameters {1}'.format(str(current),str(model.params)))\n",
    "                vm.boolean = True\n",
    "                return vm\n",
    "\n",
    "            return vm\n",
    "        if float(ampl) in vm.lookup:\n",
    "            return vm\n",
    "\n",
    "    from itertools import repeat\n",
    "    import numpy as np\n",
    "    import copy\n",
    "    import pdb\n",
    "    import get_neab\n",
    "\n",
    "    def init_vm(vm):\n",
    "        if vm.initiated == True:\n",
    "            # expand values in the range to accomodate for mutation.\n",
    "            # but otherwise exploit memory of this range.\n",
    "\n",
    "            if type(vm.steps) is type(float):\n",
    "                vm.steps = [ 0.75 * vm.steps, 1.25 * vm.steps ]\n",
    "            elif type(vm.steps) is type(list):\n",
    "                vm.steps = [ s * 1.25 for s in vm.steps ]\n",
    "            #assert len(vm.steps) > 1\n",
    "            vm.initiated = True # logically unnecessary but included for readibility\n",
    "\n",
    "        if vm.initiated == False:\n",
    "            import quantities as pq\n",
    "            import numpy as np\n",
    "            vm.boolean = False\n",
    "            steps = np.linspace(0,250,7.0)\n",
    "            steps_current = [ i*pq.pA for i in steps ]\n",
    "            vm.steps = steps_current\n",
    "            vm.initiated = True\n",
    "        return vm\n",
    "\n",
    "    def find_rheobase(vm):\n",
    "        from neuronunit.models import backends\n",
    "        from neuronunit.models.reduced import ReducedModel\n",
    "        import get_neab\n",
    "        #print(pid_map[int(os.getpid())])\n",
    "\n",
    "        new_file_path = str(get_neab.LEMS_MODEL_PATH)+str(os.getpid())\n",
    "        model = ReducedModel(new_file_path,name=str('vanilla'),backend='NEURON')\n",
    "        model.load_model()\n",
    "        model.update_run_params(vm.attrs)\n",
    "        cnt = 0\n",
    "        # If this it not the first pass/ first generation\n",
    "        # then assume the rheobase value found before mutation still holds until proven otherwise.\n",
    "        if type(vm.rheobase) is not type(None):\n",
    "            vm = check_current(vm.rheobase,vm)\n",
    "        # If its not true enter a search, with ranges informed by memory\n",
    "        cnt = 0\n",
    "        while vm.boolean == False:\n",
    "            for step in vm.steps:\n",
    "                vm = check_current(step, vm)\n",
    "                vm = check_fix_range(vm)\n",
    "                cnt+=1\n",
    "                print(cnt)\n",
    "        return vm\n",
    "\n",
    "    ## initialize where necessary.\n",
    "    #import time\n",
    "    vmpop = list(dview.map_sync(init_vm,vmpop))\n",
    "\n",
    "    # if a population has already been evaluated it may be faster to let it\n",
    "    # keep its previous rheobase searching range where this\n",
    "    # memory of a previous range as acts as a guess as the next mutations range.\n",
    "\n",
    "    vmpop = list(dview.map_sync(find_rheobase,vmpop))\n",
    "\n",
    "    return vmpop, pop\n",
    "\n",
    "\n",
    "class Individual(object):\n",
    "     '''\n",
    "     When instanced the object from this class is used as one unit of chromosome or allele by DEAP.\n",
    "     Extends list via polymorphism.\n",
    "     '''\n",
    "     def __init__(self, *args):\n",
    "         list.__init__(self, *args)\n",
    "         self.error=None\n",
    "         self.results=None\n",
    "         self.name=''\n",
    "         self.attrs = {}\n",
    "         self.params=None\n",
    "         self.score=None\n",
    "         self.fitness=None\n",
    "         self.lookup={}\n",
    "         self.rheobase=None\n",
    "         self.fitness = creator.FitnessMin\n",
    "\n",
    "creator.create(\"FitnessMin\", base.Fitness, weights=(-1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0))\n",
    "creator.create(\"Individual\", list, fitness=creator.FitnessMin)\n",
    "import os\n",
    "def get_trans_dict(param_dict):\n",
    "    trans_dict = {}\n",
    "    for i,k in enumerate(list(param_dict.keys())):\n",
    "        trans_dict[i]=k\n",
    "    return trans_dict\n",
    "import model_parameters\n",
    "param_dict = model_parameters.model_params\n",
    "td = get_trans_dict(param_dict)\n",
    "\n",
    "os.system('sudo /opt/conda/bin/pip install --upgrade networkx')\n",
    "\n",
    "cd = pickle.load(open('complete_dump.p','rb'))\n",
    "print(len(cd))\n",
    "\n",
    "unev = pickle.load(open('un_evolved.p','rb'))\n",
    "\n",
    "#unpack = [vmoffspring,history,logbook,rheobase_values,best_worst,vmhistory,hvolumes ]\n",
    "vmoffspring,history,logbook,rheobase_values,best_worst,vmhistory,hvolumes = cd[0], cd[1], cd[2], cd[3], cd[4], cd[5], cd[6]\n",
    "\n",
    "import net_graph\n",
    "\n",
    "unev = pickle.load(open('un_evolved.p','rb'))\n",
    "unev, rh_values_unevolved = unev[0], unev[1]\n",
    "\n",
    "for x,y in enumerate(vmoffspring):\n",
    "    y.rheobase = rheobase_values[x]\n",
    "    print(y,y.rheobase)\n",
    "best = vmoffspring[0]\n",
    "worst = vmoffspring[-1]\n",
    "\n",
    "assert type(best.rheobase) is not type(None)\n",
    "\n",
    "#fig, ax, plt = \n",
    "#net_graph.sp_like_shadow([best,worst])\n",
    "net_graph.sp_like_shadow(unev[4:-1])\n",
    "\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df, threed, columns1 , stacked, html = net_graph.bar_chart(best)\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "best_ = best_worst[0]\n",
    "worst_ = best_worst[1]\n",
    "\n",
    "net_graph.plot_log(logbook,hvolumes)\n",
    "net_graph.plot_objectives_history(logbook)\n",
    "df, threed, columns1 , stacked,html = net_graph.bar_chart(best)\n",
    "\n",
    "net_graph.not_just_mean(logbook,hvolumes)\n",
    "\n",
    "for x,y in enumerate(unev):\n",
    "    y.rheobase = rh_values_unevolved[x]\n",
    "vmoffspring.extend(unev)\n",
    "\n",
    "net_graph.shadow(vmoffspring,best)\n",
    "net_graph.plot_evaluate(best,worst,names=['best','worst'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "best_worst, _ = check_rheobase(best_worst)\n",
    "best_worst[0].rheobase\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_worst[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import get_neab\n",
    "get_neab.tests[4].observation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
