{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan\n",
      "/home/jovyan/mnt\n",
      "hello from before cpu \n",
      "[0, 1, 2, 3, 4, 5, 6, 7]\n",
      "[0, 1, 1024, 59049, 1048576, 9765625, 60466176, 282475249, 1073741824, 3486784401, 10000000000, 25937424601, 61917364224, 137858491849, 289254654976, 576650390625, 1099511627776, 2015993900449, 3570467226624, 6131066257801, 10240000000000, 16679880978201, 26559922791424, 41426511213649, 63403380965376, 95367431640625, 141167095653376, 205891132094649, 296196766695424, 420707233300201, 590490000000000, 819628286980801]\n",
      "[0, 1, 1024, 59049, 1048576, 9765625, 60466176, 282475249, 1073741824, 3486784401, 10000000000, 25937424601, 61917364224, 137858491849, 289254654976, 576650390625, 1099511627776, 2015993900449, 3570467226624, 6131066257801, 10240000000000, 16679880978201, 26559922791424, 41426511213649, 63403380965376, 95367431640625, 141167095653376, 205891132094649, 296196766695424, 420707233300201, 590490000000000, 819628286980801] parallel_reult\n"
     ]
    }
   ],
   "source": [
    "import os;\n",
    "%cd ~\n",
    "%cd mnt\n",
    "%pwd\n",
    "\n",
    "import os\n",
    "os.system('ipcluster start -n 8 --engines=MPIEngineSetLauncher --profile=chase --debug &')\n",
    "\n",
    "#os.system('sleep 15 &')\n",
    "import ipyparallel as ipp\n",
    "rc = ipp.Client(profile='chase');\n",
    "rc[:].use_cloudpickle()\n",
    "\n",
    "print('hello from before cpu ');\n",
    "print(rc.ids)\n",
    "dview = rc[:]\n",
    "\n",
    "serial_result = list(map(lambda x:x**10, range(32)))\n",
    "parallel_result = list(dview.map(lambda x: x**10, range(32)))\n",
    "print(serial_result)\n",
    "print(parallel_result, 'parallel_reult')\n",
    "assert serial_result == parallel_result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "attempting to recover from pickled file\n",
      "<module 'model_parameters' from '/home/jovyan/mnt/model_parameters.py'>\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "Mechanisms already loaded from path: /home/jovyan/mnt/NeuroML2.  Aborting.\n",
      "Ignoring included LEMS file: Cells.xml\n",
      "Ignoring included LEMS file: Networks.xml\n",
      "Ignoring included LEMS file: Simulation.xml\n",
      "<bound method View.map_sync of <DirectView [0, 1, 2, 3,...]>>\n",
      "<class 'method'>\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "import matplotlib as mpl\n",
    "mpl.use('agg',warn=False)\n",
    "from matplotlib import pyplot as plt\n",
    "import time\n",
    "import pdb\n",
    "import array\n",
    "import random\n",
    "import sys\n",
    "sys.path.insert(0,\"../../../sciunit\")\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import quantities as pq\n",
    "from deap import algorithms\n",
    "from deap import base\n",
    "from deap.benchmarks.tools import diversity, convergence, hypervolume\n",
    "from deap import creator\n",
    "from deap import tools\n",
    "\n",
    "import scoop\n",
    "\n",
    "try:\n",
    "    import get_neab\n",
    "except ImportError:\n",
    "    from neuronunit.tests import get_neab\n",
    "\n",
    "import quantities as qt\n",
    "import os\n",
    "import os.path\n",
    "from scoop import utils\n",
    "from scoop import futures\n",
    "\n",
    "import sciunit.scores as scores\n",
    "history = tools.History()\n",
    "\n",
    "\n",
    "\n",
    "init_start=time.time()\n",
    "creator.create(\"FitnessMin\", base.Fitness, weights=(-1.0, -1.0, -1.0, -1.0,\n",
    "                                                    -1.0, -1.0, -1.0, -1.0))\n",
    "creator.create(\"Individual\",list, fitness=creator.FitnessMin)\n",
    "\n",
    "class Individual(object):\n",
    "    '''\n",
    "    When instanced the object from this class is used as one unit of chromosome or allele by DEAP.\n",
    "    Extends list via polymorphism.\n",
    "    '''\n",
    "    def __init__(self, *args):\n",
    "        list.__init__(self, *args)\n",
    "        self.error=None\n",
    "        self.results=None\n",
    "        self.name=''\n",
    "        self.attrs = {}\n",
    "        self.params=None\n",
    "        self.score=None\n",
    "        self.fitness=None\n",
    "        self.s_html=None\n",
    "        self.lookup={}\n",
    "        self.rheobase=None\n",
    "toolbox = base.Toolbox()\n",
    "try:\n",
    "    #import model_parameters as modelp\n",
    "    import neuronunit.tests.model_parameters as modelp\n",
    "\n",
    "except ImportError:\n",
    "    from neuronunit.tests import model_parameters as params\n",
    "import model_parameters as modelp\n",
    "#import os\n",
    "#os.getcwd()\n",
    "print(modelp)\n",
    "BOUND_LOW = [ np.min(i) for i in modelp.model_params.values() ]\n",
    "BOUND_UP = [ np.max(i) for i in modelp.model_params.values() ]\n",
    "\n",
    "NDIM = len(BOUND_UP) #= len(BOUND_LOW)\n",
    "\n",
    "import functools\n",
    "\n",
    "def uniform(low, up, size=None):\n",
    "    try:\n",
    "        return [random.uniform(a, b) for a, b in zip(low, up)]\n",
    "    except TypeError:\n",
    "        return [random.uniform(a, b) for a, b in zip([low] * size, [up] * size)]\n",
    "\n",
    "toolbox.register(\"attr_float\", uniform, BOUND_LOW, BOUND_UP, NDIM)\n",
    "toolbox.register(\"Individual\", tools.initIterate, creator.Individual, toolbox.attr_float)\n",
    "import deap as deap\n",
    "\n",
    "toolbox.register(\"population\", tools.initRepeat, list, toolbox.Individual)\n",
    "toolbox.register(\"select\", tools.selNSGA2)\n",
    "\n",
    "import utilities as outils\n",
    "model = outils.model\n",
    "#outils.futures.map = dview.map_sync\n",
    "outils.map = outils.set_map(dview.map_sync)\n",
    "map = outils.map\n",
    "print(outils.map)\n",
    "print(type(outils.map))\n",
    "import quantities as pq\n",
    "import neuronunit.capabilities as cap\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def evaluate_e(individual,tuple_params):#This method must be pickle-able for scoop to work.\n",
    "    '''\n",
    "    Inputs: An individual gene from the population that has compound parameters, and a tuple iterator that\n",
    "    is a virtual model object containing an appropriate parameter set, zipped togethor with an appropriate rheobase\n",
    "    value, that was found in a previous rheobase search.\n",
    "\n",
    "    outputs: a tuple that is a compound error function that NSGA can act on.\n",
    "\n",
    "    Assumes rheobase for each individual virtual model object (vms) has already been found\n",
    "    there should be a check for vms.rheobase, and if not then error.\n",
    "    Inputs a gene and a virtual model object.\n",
    "    outputs are error components.\n",
    "    '''\n",
    "    gen,vms,rheobase=tuple_params\n",
    "    assert vms.rheobase == rheobase\n",
    "    print('{0}'.format(type(vms.rheobase)))\n",
    "    try:\n",
    "        assert type(vms.rheobase) is not type(None)\n",
    "\n",
    "        params = outils.params\n",
    "        model = outils.model\n",
    "        print(rheobase,model,params,tuple_params)\n",
    "        #pdb.set_trace()\n",
    "        uc = {'amplitude':vms.rheobase}\n",
    "        current = params.copy()['injected_square_current']\n",
    "        current.update(uc)\n",
    "        current = {'injected_square_current':current}\n",
    "        #Its very important to reset the model here. Such that its vm is new, and does not carry charge from the last simulation\n",
    "        model.update_run_params(vms.attrs)\n",
    "        model.inject_square_current(current)\n",
    "        #reset model, clear away charge from previous model\n",
    "        model.update_run_params(vms.attrs)\n",
    "        n_spikes = model.get_spike_count()\n",
    "        sane = get_neab.suite.tests[0].sanity_check(vms.rheobase*pq.pA,model)\n",
    "\n",
    "        try:\n",
    "            assert n_spikes == 1 and sane == True #or n_spikes == 0  # Its possible that no rheobase was found\n",
    "            for i in [4,5,6]:\n",
    "                get_neab.suite.tests[i].params['injected_square_current']['amplitude'] = vms.rheobase*pq.pA\n",
    "            get_neab.suite.tests[0].prediction={}\n",
    "            assert type(vms.rheobase) != type(None)\n",
    "            score = get_neab.suite.tests[0].prediction['value']=vms.rheobase * pq.pA\n",
    "            score = get_neab.suite.judge(model)#passing in model, changes the model\n",
    "            model.run_number+=1\n",
    "            error = score.sort_key.values.tolist()[0]\n",
    "            for i in error:\n",
    "                if type(i) is type(None):\n",
    "                    i = np.mean(error)\n",
    "            #individual.error = error\n",
    "            individual.rheobase = vms.rheobase\n",
    "        except:\n",
    "\n",
    "            inderr = getattr(individual, \"error\", None)\n",
    "            if type(inderr) is not (None):\n",
    "                if len(individual.error)!=0:\n",
    "                    #the average of 10 and the previous score is chosen as a nominally high distance from zero\n",
    "                    error = [ (abs(-10.0+i)/2.0) for i in individual.error ]\n",
    "            else:\n",
    "                error = [ 100.0 for i in range(0,8) ]\n",
    "\n",
    "    except Exception as e:\n",
    "        #raise type(vms.rheobase) is type(None)\n",
    "        error = [ 100.0 for i in range(0,8) ]\n",
    "\n",
    "    return error[0],error[1],error[2],error[3],error[4],error[5],error[6],error[7],\n",
    "\n",
    "#param_dict should be one of scoops shared constants\n",
    "\n",
    "\n",
    "toolbox.register(\"evaluate\", evaluate_e)\n",
    "toolbox.register(\"mate\", tools.cxSimulatedBinaryBounded, low=BOUND_LOW, up=BOUND_UP, eta=20.0)\n",
    "toolbox.register(\"mutate\", tools.mutPolynomialBounded, low=BOUND_LOW, up=BOUND_UP, eta=20.0, indpb=1.0/NDIM)\n",
    "\n",
    "toolbox.decorate(\"mate\", history.decorator)\n",
    "toolbox.decorate(\"mutate\", history.decorator)\n",
    "\n",
    "toolbox.register(\"select\", tools.selNSGA2)\n",
    "toolbox.register(\"map\", dview.map_sync)\n",
    "\n",
    "\n",
    "def plot_ss(vmlist,gen):\n",
    "    '''\n",
    "    '''\n",
    "\n",
    "    import matplotlib.pyplot as plt\n",
    "    plt.clf()\n",
    "    for ind,j in enumerate(vmlist):\n",
    "        if hasattr(ind,'results'):\n",
    "            plt.plot(ind.results['t'],ind.results['vm'],label=str(vmlist[j].attr) )\n",
    "            #plt.xlabel(str(vmlist[j].attr))\n",
    "    plt.savefig('snap_shot_at_gen_'+str(gen)+'.png')\n",
    "    plt.clf()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def get_trans_dict(param_dict):\n",
    "\n",
    "    trans_dict = {}\n",
    "    for i,k in enumerate(list(param_dict.keys())):\n",
    "        trans_dict[i]=k\n",
    "    return trans_dict\n",
    "\n",
    "import model_parameters\n",
    "\n",
    "param_dict = model_parameters.model_params\n",
    "#global trans_dict, param_dict\n",
    "\n",
    "def individual_to_vm(ind,trans_dict=None):\n",
    "    '''\n",
    "    This method still not a reliable mapping\n",
    "    Creates vm.attrs from param_dict and a\n",
    "    translation dictionary\n",
    "    '''\n",
    "    vm = outils.VirtualModel()\n",
    "    param_dict={}\n",
    "    #if type(trans_dict) is not type(None):\n",
    "    if type(trans_dict) is not type(None):\n",
    "\n",
    "        'transdict is: {0} ind is {1}'.format(trans_dict, ind)\n",
    "        for i, j in enumerate(ind):\n",
    "            print(i,j)\n",
    "            param_dict[trans_dict[i]]=str(j)\n",
    "        vm.attrs = param_dict\n",
    "        vm.trans_dict = trans_dict\n",
    "    #return vm\n",
    "\n",
    "    if type(trans_dict) is type(None):\n",
    "        for i, j in enumerate(ind):\n",
    "            vm.attrs[vm.trans_dict[i]]=str(j)\n",
    "    return vm\n",
    "\n",
    "\n",
    "def replace_rh(pop,vmpop):\n",
    "    '''\n",
    "    #discard models that cause None rheobase results,\n",
    "    # and create new models by mutating away from the corresponding  parameters.\n",
    "    #make sure that the old individual, and virtual model object are\n",
    "    #over written so do not use list append pattern, as this will not\n",
    "    #over write the objects in place, but instead just grow the lists inappropriately\n",
    "    #also some of the lines below may be superflous in terms of machine instructions, but\n",
    "    #they function to make the code more explicit and human readable.\n",
    "    '''\n",
    "    rheobase_checking=outils.rheobase_checking\n",
    "    from itertools import repeat\n",
    "    import copy\n",
    "    for i,ind in enumerate(pop):\n",
    "        j=0\n",
    "        while type(vmpop[i].rheobase) is type(None):\n",
    "            print(j)\n",
    "            j+=1\n",
    "            #print('this loop appropriately exits none mutate away from ')\n",
    "            toolbox.mutate(ind)\n",
    "            toolbox.mutate(ind)\n",
    "            toolbox.mutate(ind)\n",
    "            print('trying mutations: {0}'.format(ind))\n",
    "            #temp = individual_to_vm(ind,param_dict)\n",
    "            trans_dict=vmpop[i].trans_dict\n",
    "            vm_temp = individual_to_vm(ind,trans_dict)\n",
    "            vmpop[i] = rheobase_checking(vm_temp)\n",
    "            'trying value {0}'.format(vmpop[i].rheobase)\n",
    "            ind.rheobase = vmpop[i].rheobase\n",
    "            pop[i] = ind\n",
    "\n",
    "            #if type(vmpop[i].rheobase) is not type(None):\n",
    "            print('rheobase value is updating {0}'.format(vmpop[i].rheobase))\n",
    "            if type(vmpop[i].rheobase) is not type(None):\n",
    "                break\n",
    "        assert type(vmpop[i].rheobase) is not type(None)\n",
    "    assert ind.rheobase == vmpop[i].rheobase\n",
    "    assert len(pop)!=0\n",
    "    assert len(vmpop)!=0\n",
    "\n",
    "\n",
    "    return pop, vmpop\n",
    "\n",
    "\n",
    "\n",
    "def test_to_model(vms,local_test_methods):\n",
    "    import get_neab\n",
    "    tests = get_neab.suite.tests\n",
    "    import matplotlib.pyplot as plt\n",
    "    import copy\n",
    "    global model\n",
    "    model.update_run_params(vms.attrs)\n",
    "    tests = None\n",
    "    tests = get_neab.suite.tests\n",
    "    tests[0].prediction={}\n",
    "    tests[0].prediction['value']=vms.rheobase*qt.pA\n",
    "    tests[0].params['injected_square_current']['amplitude']=vms.rheobase*qt.pA\n",
    "    #TODO all of the external rheobase related things need to be re-encapsulated into the NeuroUnit class.\n",
    "    if local_test_methods in [4,5,6]:\n",
    "        tests[local_test_methods].params['injected_square_current']['amplitude']=vms.rheobase*qt.pA\n",
    "    #model.results['vm'] = [ 0 ]\n",
    "    model.re_init(vms.attrs)\n",
    "    tests[local_test_methods].generate_prediction(model)\n",
    "\n",
    "\n",
    "\n",
    "def update_vm_pop(pop,trans_dict,rh_value=None):\n",
    "    '''\n",
    "    inputs a population of genes/alleles, the population size MU, and an optional argument of a rheobase value guess\n",
    "    outputs a population of genes/alleles, a population of individual object shells, ie a pickleable container for gene attributes.\n",
    "    Rationale, not every gene value will result in a model for which rheobase is found, in which case that gene is discarded, however to\n",
    "    compensate for losses in gene population size, more gene samples must be tested for a successful return from a rheobase search.\n",
    "    If the tests return are successful these new sampled individuals are appended to the population, and then their attributes are mapped onto\n",
    "    corresponding virtual model objects.\n",
    "    '''\n",
    "    from itertools import repeat\n",
    "    import numpy as np\n",
    "    import copy\n",
    "    import pdb\n",
    "    rheobase_checking=outils.rheobase_checking\n",
    "    vmpop = list(dview.map_sync(individual_to_vm,[toolbox.clone(i) for i in copy.copy(pop)], repeat(trans_dict) ))\n",
    "\n",
    "    #vmpop = list(futures.map(individual_to_vm,[toolbox.clone(i) for i in copy.copy(pop)], repeat(trans_dict) ))\n",
    "    vmpop = list(                                                                                                                                   dview.map_sync(rheobase_checking,vmpop))\n",
    "    print('checkpoint 1 output from parallel map {0}'.format(vmpop))\n",
    "    rh_value = [ i.rheobase for i in vmpop]\n",
    "    print(rh_value)\n",
    "\n",
    "    def rbc(rh_value):\n",
    "        boolean_r_check=False\n",
    "        for r in rh_value:\n",
    "            print(r)\n",
    "            if type(r) is type(None):\n",
    "                print(type(r))\n",
    "                boolean_r_check == True\n",
    "        return boolean_r_check\n",
    "\n",
    "    while rbc(rh_value) is True:\n",
    "        pop,vmpop = replace_rh(pop,vmpop,rh_value)\n",
    "        rh_value = [ i.rheobase for i in vmpop ]\n",
    "\n",
    "    rh_value = [ toolbox.clone(i).rheobase for i in copy.copy(vmpop) ]\n",
    "\n",
    "\n",
    "    assert len(pop) == len(vmpop)\n",
    "    assert len(pop)!=0\n",
    "    assert len(vmpop)!=0\n",
    "    assert rbc(rh_value) is False\n",
    "    for y,x in enumerate(vmpop):\n",
    "        assert x.rheobase== rh_value[y]\n",
    "        print(x.rheobase)\n",
    "        assert type(x.rheobase) is not type(None)\n",
    "    return pop,vmpop,rh_value\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "from scoop import futures, _control, utils, shared\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "def main():\n",
    "    global NGEN\n",
    "    NGEN=3\n",
    "    global MU\n",
    "    import numpy as np\n",
    "    #MU=8#Mu must be some multiple of 4, such that it can be split into even numbers over 8 CPUs\n",
    "    MU=8\n",
    "    CXPB = 0.9\n",
    "    #stats = tools.Statistics(lambda ind: ind.fitness.values)\n",
    "    pf = tools.ParetoFront()\n",
    "    from scoop.fallbacks import NotStartedProperly\n",
    "    trans_dict = get_trans_dict(param_dict)\n",
    "    if scoop.fallbacks.NotStartedProperly()==False:\n",
    "        print('this never evaluates {0}'.format(scoop.fallbacks.NotStartedProperly()))\n",
    "\n",
    "        import pdb; pdb.set_trace()\n",
    "        shared.setConst(td = trans_dict)\n",
    "        td = shared.getConst('td')\n",
    "        print('the shared constant {0}'.format(shared.getConst('td')))\n",
    "    else:\n",
    "        td = trans_dict\n",
    "    pop = toolbox.population(n = MU)\n",
    "    pop = [toolbox.clone(i) for i in pop]\n",
    "    #history.update(pop)\n",
    "    pf.update([toolbox.clone(i) for i in pop])\n",
    "    #Now attempt to get the rheobase values by first trying the mean rheobase value.\n",
    "    #This is not an exhaustive search that results in found all rheobase values\n",
    "    #It is just a trying out an educated guess on each individual in the whole population as a first pass.\n",
    "    #invalid_ind = [ ind for ind in pop if not ind.fitness.valid ]\n",
    "\n",
    "    #rheobase_checking = outils.rheobase_checking\n",
    "\n",
    "    pop,vmpop,rhstorage = list(update_vm_pop(pop,td))\n",
    "    for y,x in enumerate(vmpop):\n",
    "        assert x.rheobase== rhstorage[y]\n",
    "        print(x.rheobase,rhstorage[y])\n",
    "        #assert type(x.rheobase) is not type(None)\n",
    "    assert len(pop) != 0\n",
    "    print(type(pop),type(vmpop))\n",
    "    assert len(pop) == len(vmpop)\n",
    "    assert len(vmpop) != 0\n",
    "    assert len(pop) != 0\n",
    "    for i in vmpop:\n",
    "        i.td=shared.getConst('td')\n",
    "    print('updatevmpop returns a whole heap of nones suggesting its not working {0}'.format(vmpop))\n",
    "    #population may also be altered in this process.\n",
    "    pf.update(pop)\n",
    "    #rhstorage = [ item.rheobase for item in vmpop]\n",
    "\n",
    "    from itertools import repeat\n",
    "    #repeat 0, for generation 0\n",
    "    tuple_storage = zip(repeat(0),vmpop,rhstorage)\n",
    "    #Now get the fitness of genes:\n",
    "    #Note the evaluate function called is different\n",
    "    #pop is being passed attributes or it needs to create them inside evaluate.\n",
    "\n",
    "    fitnesses = list(dview.map_sync(toolbox.evaluate, pop, tuple_storage))\n",
    "    invalid_ind = [ ind for ind in pop if not ind.fitness.valid ]\n",
    "    return fitnesses,invalid_ind\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fitnesses,invalid_ind = main()\n",
    "print(fitness,invalid_ind)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cd mnt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cd neuronunit/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cd neuronunit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cd tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
